Use pod Deployment name as workload name for flow workload field

chancez opened the pull request:
In #16514 support for retrieving workload names and storing them in flows was added for L3/L4 flows. In #21039 this same functionality has been added to L7 flows. However, currently for pods owned by ReplicaSets, this is somewhat limited, as most usages of ReplicaSets are via Deployments, and most users would want the Deployment name for the workload, rather than the ReplicaSet names. 

This is particularly important for metrics, where you would want to be able to identify the source/destination workload or filter/aggregate by the workload name, which is common in other service meshes.

WIth this PR, we extend the previous functionality to now determine the Deployment name from the pod. This functionality was taken from https://github.com/istio/istio/blob/1aca7a67afd7b3e1d24fafb2fbfbeaf1e41534c0/pkg/kube/util.go and adapted to work with the cilium slim API types.

This approach reverses the way pod names are generated for Deployments (and a few other types such as ReplicationControllers, CronJobs, and DeploymentConfigs), meaning we can do this without adding more k8s watchers and caches for additional Kubernetes resources, which means less memory consumption. 

```release-note
Use pod Deployment name as workload name for flow workload field
```

chancez committed:
k8s/slim: Add GenerateName to ObjectMeta and Controller to OwnerReference

These are needed to heuristically lookup the workload owner for a pod
based on the ownerReference of a pod.

Signed-off-by: Chance Zibolski <chance.zibolski@gmail.com>

chancez committed:
k8s/utils: Add GetWorkloadMetaFromPod from Istio

Vendor a modified version of Istio's GetDeployMetaFromPod from
github.com/istio/istio/pkg/kube/util.go as GetWorkloadMetaFromPod. It has
been modified to work with the cilium slim API types and renamed to use
the workload terminology.

Signed-off-by: Chance Zibolski <chance.zibolski@gmail.com>

michi-covalent reviewed (approved):
lgtm ✅ maybe add some unit tests for `GetWorkloadMetaFromPod`

chancez committed:
hubble/parser: Use Deployment name in flow workload

In #16514 support for retrieving workload names and storing them in
flows was added for L3/L4 flows. In #21039 this same functionality has
been added to L7 flows. However, currently for pods owned by
ReplicaSets, this is somewhat limited, as most usages of ReplicaSets are
via Deployments, and most users would want the Deployment name for the
workload, rather than the ReplicaSet names.

This is particularly important for metrics, where you would want to be
able to identify the source/destination workload or filter/aggregate by
the workload name, which is common in other service meshes.

With this commit, we extend the previous functionality to now determine
the Deployment name from the pod.

This functionality was taken from
https://github.com/istio/istio/blob/1aca7a67afd7b3e1d24fafb2fbfbeaf1e41534c0/pkg/kube/util.go
and adapted to work with the cilium slim API types.

This approach reverses the way pod names are generated for Deployments
(and a few other types such as ReplicationControllers, CronJobs, and
DeploymentConfigs), meaning we can do this without adding more k8s
watchers and caches for additional Kubernetes resources, which means
less memory consumption.

Signed-off-by: Chance Zibolski <chance.zibolski@gmail.com>

christarazi commented on pkg/hubble/parser/seven/parser.go:
```diff
@@ -176,10 +177,9 @@ func (p *Parser) computeResponseTime(r *accesslog.LogRecord, timestamp time.Time
-				endpoint.Workloads[index] = &pb.Workload{Kind: owner.Kind, Name: owner.Name}
+			workload, workloadTypeMeta, ok := utils.GetWorkloadMetaFromPod(pod)
+			if ok {
+				endpoint.Workloads = []*pb.Workload{{Kind: workloadTypeMeta.Kind, Name: workload.Name}}
```
Could this be an `append` to the slice?

chancez replied:
This is the only place it's used besides the other parser, so is that more style or for ensuring if something else ends up setting this field, we're not overwriting it?

christarazi replied:
More for the latter, but if it's not being used elsewhere, then maybe `append` is not necessary. `append` is just an approach which doesn't assume the state of the field.

chancez replied:
I think I'll keep it how it is now then. If we end up needing to source this from multiple places, then I suspect just appending won't necessarily be correct either, since we may end up with duplicates, etc.

tommyp1ckles commented on pkg/k8s/utils/workload.go:
```diff
@@ -0,0 +1,93 @@
+			workloadObjectMeta.Name = controllerRef.Name
+
+			// heuristic for deployment detection
+			if typeMetadata.Kind == "ReplicaSet" && pod.Labels["pod-template-hash"] != "" && strings.HasSuffix(controllerRef.Name, pod.Labels["pod-template-hash"]) {
```
Would it be worth checking if pod-template-hash is in the map, or should that never happen?

chancez replied:
It would happen if using a ReplicaSet that's not managed by a Deployment. It shouldn't matter though, since the the zero value is an empty string. The Deployment tests cover this case.

tommyp1ckles commented on pkg/k8s/utils/workload.go:
```diff
@@ -0,0 +1,93 @@
+			} else if typeMetadata.Kind == "ReplicaSet" && pod.Labels["pod-template-hash"] == "" {
+				workloadObjectMeta.Name = controllerRef.Name
+				typeMetadata.Kind = "ReplicaSet"
+			} else if typeMetadata.Kind == "ReplicationController" && pod.Labels["deploymentconfig"] != "" {
```
Same question as above, but for deploymentconfig here.

chancez replied:
See above

tommyp1ckles commented on pkg/hubble/parser/seven/parser.go:
```diff
@@ -176,10 +177,9 @@ func (p *Parser) computeResponseTime(r *accesslog.LogRecord, timestamp time.Time
-			endpoint.Workloads = make([]*pb.Workload, olen)
-			for index, owner := range pod.GetOwnerReferences() {
-				endpoint.Workloads[index] = &pb.Workload{Kind: owner.Kind, Name: owner.Name}
+			workload, workloadTypeMeta, ok := utils.GetWorkloadMetaFromPod(pod)
```
What happens if the pod belongs to a StatefulSet?

chancez replied:
I'll add a test for that. It should fallback to the controller ref, which for a pod managed by a StatefulSet is the StatefulSet.

chancez replied:
Added a new test in the latest changes. ✔️

tommyp1ckles reviewed (commented):
Looks good, just had a few small questions.

kaworu reviewed (approved):
Thanks @chancez patch LGTM.

@qmonnet Could you please take a look at the copyright attribution? (see [this comment](https://github.com/cilium/cilium/pull/21124#issuecomment-1230972090))? :pray:

qmonnet commented on pkg/k8s/utils/workload.go:
```diff
@@ -0,0 +1,93 @@
+// https://github.com/istio/istio/blob/1aca7a67afd7b3e1d24fafb2fbfbeaf1e41534c0/pkg/kube/util.go
+//
+// Modifications:
+// GetDeployMetaFromPod has been renamed to to GetWorkloadMetaFromPod and has
```
```suggestion
// GetDeployMetaFromPod has been renamed to GetWorkloadMetaFromPod and has
```

Also, typo in commit log (`istio/isto/...`)

chancez replied:
Ah good catch. Fixed the typos.

qmonnet reviewed (approved):
> @qmonnet Could you please take a look at the copyright attribution? (see [this comment](https://github.com/cilium/cilium/pull/21124#issuecomment-1230972090))? pray

“I'm not a lawyer”, but looks good as far as I can tell. Kept the original copyright and location, same license, mentioned the differences - looks all good to me.
